src/backend/distributed/commands/function.c

```diff
@@ -1303,18 +1286,116 @@ PostprocessCreateFunctionStmt(Node *node, const char *queryString)
-								GetFunctionAlterOwnerCommand(address.objectId),
-								ENABLE_DDL_PROPAGATION);
+	if (IsObjectAddressOwnedByExtension(&functionAddress, NULL))
+	{
+		return NIL;
+	}
+
+	/*
+	 * This check should have been valid for all objects not only for functions. Though,
+	 * we do this limited check for now as functions are more likely to be used with
+	 * such dependencies, and we want to scope it for now
+	 */
+	ObjectAddress *undistributableDependency = GetUndistributableDependency(
+		&functionAddress);
+	if (undistributableDependency != NULL)
+	{
+		if (SupportedDependencyByCitus(undistributableDependency))
+		{
+			/*
+			 * Citus can't distribute some relations as dependency, although those
+			 * types as supported by Citus. So we can use get_rel_name directly
+			 */
+			RangeVar *functionRangeVar = makeRangeVarFromNameList(stmt->funcname);
+			char *functionName = functionRangeVar->relname;
+			char *dependentRelationName =
+				get_rel_name(undistributableDependency->objectId);
+
+			ereport(WARNING, (errmsg("Citus can't distribute function \"%s\" having "
+									 "dependency on non-distributed relation \"%s\"",
+									 functionName, dependentRelationName),
+							  errdetail("Function will be created only locally"),
+							  errhint("To distribute function, distribute dependent "
+									  "relations first. Then, re-create the function")));
+		}
+		else
+		{
+			char *objectType = NULL;
+			#if PG_VERSION_NUM >= PG_VERSION_14
+			objectType = getObjectTypeDescription(undistributableDependency, false);
+			#else
+			objectType = getObjectTypeDescription(undistributableDependency);
+			#endif
+			ereport(WARNING, (errmsg("Citus can't distribute functions having "
+									 "dependency on unsupported object of type \"%s\"",
+									 objectType),
+							  errdetail("Function will be created only locally")));
+		}
+
+		return NIL;
+	}
+
+	EnsureDependenciesExistOnAllNodes(&functionAddress);
+
+	List *commands = list_make1(DISABLE_DDL_PROPAGATION);
+	commands = list_concat(commands, CreateFunctionDDLCommandsIdempotent(
+							   &functionAddress));
+	commands = list_concat(commands, list_make1(ENABLE_DDL_PROPAGATION));
 
 	return NodeDDLTaskList(NON_COORDINATOR_NODES, commands);
 }
 
 
+/*
+ * GetUndistributableDependency checks whether object has any non-distributable
+ * dependency. If any one found, it will be returned.
+ */
+static ObjectAddress *
+GetUndistributableDependency(ObjectAddress *objectAddress)
+{
+	List *dependencies = GetAllDependenciesForObject(objectAddress);
+	ObjectAddress *dependency = NULL;
+	foreach_ptr(dependency, dependencies)
+	{
+		if (IsObjectDistributed(dependency))
+		{
+			continue;
+		}
+
+		if (!SupportedDependencyByCitus(dependency))
+		{
+			/*
+			 * Although languages are not supported by Citus, all must be created
+			 * with the postgres itself, so skip them. Since roles should also be
+			 * handled manually with Citus community, skip them as well.
+			 */
```

Comment:
Few questions:

1) How do we get to `OCLASS_LANGUAGE`? When I check pg_language, what I see that some languages are created via extensions, so we cannot skip like this. Should this have been IsOwnedByExtension or such?

```SQL
 select * from pg_language ;
┌───────┬──────────┬──────────┬─────────┬──────────────┬───────────────┬───────────┬──────────────┬────────┐
│  oid  │ lanname  │ lanowner │ lanispl │ lanpltrusted │ lanplcallfoid │ laninline │ lanvalidator │ lanacl │
├───────┼──────────┼──────────┼─────────┼──────────────┼───────────────┼───────────┼──────────────┼────────┤
│    12 │ internal │       10 │ f       │ f            │             0 │         0 │         2246 │        │
│    13 │ c        │       10 │ f       │ f            │             0 │         0 │         2247 │        │
│    14 │ sql      │       10 │ f       │ t            │             0 │         0 │         2248 │        │
│ 14011 │ plpgsql  │       10 │ t       │ t            │         14008 │     14009 │        14010 │        │
└───────┴──────────┴──────────┴─────────┴──────────────┴───────────────┴───────────┴──────────────┴────────┘
(4 rows)
 SELECT pg_identify_object_as_address(classid, objid, objsubid), pg_identify_object_as_address(refclassid, refobjid, refobjsubid) from pg_depend WHERE objid = 14011;
┌───────────────────────────────┬────────────────────────────────────────────────────────────────────────┐
│ pg_identify_object_as_address │                     pg_identify_object_as_address                      │
├───────────────────────────────┼────────────────────────────────────────────────────────────────────────┤
│ (language,{plpgsql},{})       │ (extension,{plpgsql},{})                                               │
│ (language,{plpgsql},{})       │ (function,"{pg_catalog,plpgsql_validator}",{pg_catalog.oid})           │
│ (language,{plpgsql},{})       │ (function,"{pg_catalog,plpgsql_inline_handler}",{pg_catalog.internal}) │
│ (language,{plpgsql},{})       │ (function,"{pg_catalog,plpgsql_call_handler}",{})                      │
└───────────────────────────────┴────────────────────────────────────────────────────────────────────────┘
(4 rows)


```

2) In general, it should be safe to skip if `oid < FirstNormalObjectId` as they belong postgres. I wouldn't mind adding that check as well